pylops/signalprocessing/patch2d.py

```diff
@@ -172,104 +173,268 @@ def Patch2D(
+                taps[nwins - nwins1] = tapleftbottom
+                taps[nwins - 1] = taprightbottom
+                self.taps = np.vstack(taps).reshape(nwins0, nwins1, nwin[0], nwin[1])
+            else:
+                taps = [
+                    taplefttop,
+                    taptop,
+                    taprighttop,
+                    tapleft,
+                    tap,
+                    tapright,
+                    tapleftbottom,
+                    tapbottom,
+                    taprightbottom,
+                ]
+                self.taps = np.vstack(taps).reshape(3, 3, nwin[0], nwin[1])
+
+        # define scalings
+        if scalings is None:
+            self.scalings = [1.0] * nwins
+        else:
+            self.scalings = scalings
+
+        # check if operator is applied to all windows simultaneously
+        self.simOp = False
+        if Op.shape[1] == np.prod(dims):
+            self.simOp = True
+        self.Op = Op
+
+        super().__init__(
+            dtype=Op.dtype,
+            dims=(nwins0, nwins1, int(dims[0] // nwins0), int(dims[1] // nwins1)),
+            dimsd=dimsd,
+            clinear=False,
+            name=name,
         )
 
-    hstack = HStack(
-        [
-            Restriction(
-                (nwin[0], dimsd[1]), range(win_in, win_end), axis=1, dtype=Op.dtype
-            ).H
-            for win_in, win_end in zip(dwin1_ins, dwin1_ends)
-        ]
-    )
-    combining1 = BlockDiag([hstack] * nwins0)
+        self._register_multiplications(self.savetaper)
+
+    def _apply_taper(self, ywins, iwin0, iwin1):
+        if iwin0 == 0 and iwin1 == 0:
+            ywins[0, 0] = self.taps[0, 0] * ywins[0, 0]
+        elif iwin0 == 0 and iwin1 == self.dims[1] - 1:
+            ywins[0, -1] = self.taps[0, -1] * ywins[0, -1]
+        elif iwin0 == 0:
+            ywins[0, iwin1] = self.taps[0, 1] * ywins[0, iwin1]
+        elif iwin0 == self.dims[0] - 1 and iwin1 == 0:
+            ywins[-1, 0] = self.taps[-1, 0] * ywins[-1, 0]
+        elif iwin0 == self.dims[0] - 1 and iwin1 == self.dims[1] - 1:
+            ywins[-1, -1] = self.taps[-1, -1] * ywins[-1, -1]
+        elif iwin0 == self.dims[0] - 1:
+            ywins[-1, iwin1] = self.taps[-1, 1] * ywins[-1, iwin1]
+        elif iwin1 == 0:
+            ywins[iwin0, 0] = self.taps[1, 0] * ywins[iwin0, 0]
+        elif iwin1 == self.dims[1] - 1:
+            ywins[iwin0, -1] = self.taps[1, -1] * ywins[iwin0, -1]
+        else:
+            ywins[iwin0, iwin1] = self.taps[1, 1] * ywins[iwin0, iwin1]
+        return ywins
 
-    combining0 = HStack(
-        [
-            Restriction(dimsd, range(win_in, win_end), axis=0, dtype=Op.dtype).H
-            for win_in, win_end in zip(dwin0_ins, dwin0_ends)
+    @reshaped()
+    def _matvec_savetaper(self, x: NDArray) -> NDArray:
+        ncp = get_array_module(x)
+        if self.tapertype is not None:
+            self.taps = to_cupy_conditional(x, self.taps)
```

Comment:
I realized this function may have a "bug", the GPU array gets created on the current device, not the device of `x`.

The function should probably look like this (untested!):

```python
def to_cupy_conditional(x: npt.ArrayLike, y: npt.ArrayLike) -> NDArray:
    if deps.cupy_enabled:
        if cp.get_array_module(x) == cp and cp.get_array_module(y) == np:
            with cp.cuda.Device(x.device):
                y = cp.asarray(y)
    return y
```

Reply:
I think you are right, well spotted. I am going to open an issue and take this separetely.